tests/run_data_utils_test.py

```diff
@@ -0,0 +1,53 @@
+from batchglm.api.utils.data import constraint_system_from_star
+import pandas as pd
+import logging
+import numpy as np
+import unittest
+
+logger = logging.getLogger("batchglm")
+
+class TestConstraintSystemFromStar(unittest.TestCase):
+
+    def test_constraint_system_dict(self):
+        formula = '~0 + condition + batch'
+        sample_description = pd.DataFrame({ 'condition': [0, 1, 1, 0, 0, 1], 'batch': [1, 2, 2, 0, 0, 3] })
+        constraints = {"batch": "condition"}
+        dmat, coef_names, cmat, term_names = constraint_system_from_star(constraints, sample_description=sample_description, formula=formula)
+        assert term_names == ['condition', 'batch']
+        assert coef_names == ['condition[0]', 'condition[1]', 'batch[0]', 'batch[1]', 'batch[2]', 'batch[3]']
+        assert np.all(np.equal(cmat, np.array([[ 1.,  0.,  0.,  0.],
+            [ 0.,  1.,  0.,  0.],
+            [ 0.,  0., -1.,  0.],
+            [ 0.,  0.,  1.,  0.],
+            [ 0.,  0.,  0., -1.],
+            [ 0.,  0.,  0.,  1.]])))
+        assert np.all(np.equal(dmat, np.array([[1., 0., 0., 1., 0., 0.],
+            [0., 1., 0., 0., 1., 0.],
+            [0., 1., 0., 0., 1., 0.],
+            [1., 0., 1., 0., 0., 0.],
+            [1., 0., 1., 0., 0., 0.],
+            [0., 1., 0., 0., 0., 1.]])))
+    
+    def test_constraint_system_list(self):
+        formula = '~0 + condition + batch'
+        sample_description = pd.DataFrame({ 'condition': [0, 1, 1, 0, 0, 1], 'batch': [1, 2, 2, 0, 0, 3] })
+        constraints = ["condition[0] + condition[1] = 0"]
+        dmat, coef_names, cmat, term_names = constraint_system_from_star(constraints, sample_description=sample_description, formula=formula)
+        assert term_names == None
+        assert coef_names == ['condition[0]', 'condition[1]', 'batch[T.1]', 'batch[T.2]', 'batch[T.3]']
```

Comment:
Similarly, why is this different than the above?

Reply:
This is because your constraints are different. In the above case you supplied {'batch': 'condition'} which is internally parsed to ['batch[0]+batch[1]=0', 'batch[2]+batch[3]=0']. This is because you tell it to constrain the batches on the condition, which means that the batches for each condition must sum up to 1. Since batch 0 and 1 belong to condition 0 and 2 and 3 belong to condition 1, this is indeed what it does here.
In the below example you only tell it to constrain the conditions to sum up to one, irrespective of the batches.

Reply:
@picciama I was referring to the coefficient names.  If I change it to `"batch[T.0] + batch[T.1] = 0"` it actually doesn't even work.  There is no `batch[T.0]` to work with which is why this test is different. Ideally these would all be the same with the same result so that we can make sure we are consistent.

```python
self = <tests.run_data_utils_test.TestConstraintSystemFromStar testMethod=test_constraint_system_list>

    def test_constraint_system_list(self):
        formula = '~0 + condition + batch'
        sample_description = pd.DataFrame({ 'condition': [0, 1, 1, 0, 0, 1], 'batch': [1, 2, 2, 0, 0, 3] })
        constraints = ["batch[T.0] + batch[T.1] = 0"]
>       dmat, coef_names, cmat, term_names = constraint_system_from_star(constraints, sample_description=sample_description, formula=formula)

tests/run_data_utils_test.py:35:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/Caskroom/miniconda/base/envs/batchglm/lib/python3.8/functools.py:875: in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
batchglm/utils/data.py:277: in _
    cmat = constraint_matrix_from_string(
batchglm/utils/data.py:468: in constraint_matrix_from_string
    constraint_ls = [di.linear_constraint(x).coefs[0] for x in constraints]
batchglm/utils/data.py:468: in <listcomp>
    constraint_ls = [di.linear_constraint(x).coefs[0] for x in constraints]
.nox/tests-3-8/lib/python3.8/site-packages/patsy/design_info.py:536: in linear_constraint
    return linear_constraint(constraint_likes, self.column_names)
.nox/tests-3-8/lib/python3.8/site-packages/patsy/constraint.py:406: in linear_constraint
    tree = parse_constraint(code, variable_names)
.nox/tests-3-8/lib/python3.8/site-packages/patsy/constraint.py:240: in parse_constraint
    return infix_parse(_tokenize_constraint(string, variable_names),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

string = 'batch[T.0] + batch[T.1] = 0', variable_names = ['condition[0]', 'condition[1]', 'batch[T.1]', 'batch[T.2]', 'batch[T.3]']

    def _tokenize_constraint(string, variable_names):
        lparen_re = r"\("
        rparen_re = r"\)"
        op_re = "|".join([re.escape(op.token_type) for op in _ops])
        num_re = r"[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?"
        whitespace_re = r"\s+"

        # Prefer long matches:
        variable_names = sorted(variable_names, key=len, reverse=True)
        variable_re = "|".join([re.escape(n) for n in variable_names])

        lexicon = [
            (lparen_re, _token_maker(Token.LPAREN, string)),
            (rparen_re, _token_maker(Token.RPAREN, string)),
            (op_re, _token_maker("__OP__", string)),
            (variable_re, _token_maker("VARIABLE", string)),
            (num_re, _token_maker("NUMBER", string)),
            (whitespace_re, None),
            ]

        scanner = re.Scanner(lexicon)
        tokens, leftover = scanner.scan(string)
        if leftover:
            offset = len(string) - len(leftover)
>           raise PatsyError("unrecognized token in constraint",
                                Origin(string, offset, offset + 1))
E           patsy.PatsyError: unrecognized token in constraint
E               batch[T.0] + batch[T.1] = 0
E               ^

.nox/tests-3-8/lib/python3.8/site-packages/patsy/constraint.py:198: PatsyError
```